### PR TITLE
Enforce UTF-8 during wheel builds

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
Add logic to the wheel builder to set the current locale to a UTF-8 locale. This should ensure that any IPC happens in UTF-8 as opposed to just hoping that the system locale is UTF-8. Also, modify some file I/O to always specify the intended encoding.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20640)

### CODE REVIEW

**Comment by mwoehlke-kitware** [comment]
+@jwnimmer-tri for feature review, please (no rush).

This should work, though there is some argument for `LC_ALL=C LC_CTYPE=UTF-8`. (macOS unfortunately doesn't like `LC_ALL=C.UTF-8`. See also https://unix.stackexchange.com/questions/573886.)

**Comment by mwoehlke-kitware** [comment]
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-release please
@drake-jenkins-bot mac-arm-ventura-unprovisioned-clang-wheel-experimental-release please
@drake-jenkins-bot mac-x86-monterey-unprovisioned-clang-wheel-experimental-release please

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: both +(status: single reviewer ok).

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-:-NqsNIH467DjGsYZpNcI:b-fl40af)* status: 2 unresolved discussions

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMAkL5HAthJllQLCf:-NqsMAkL5HAthJllQLCg:b-ewawt2) (no related file):*
Per https://peps.python.org/pep-0686/ and my related experience, the canonical string for `encoding=` in Python should be `"utf-8"` not `"utf8"`.  Please adjust throughout the PR.

___
___
*[`tools/wheel/image/setup.py` line 46 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMX7N-_aZMrfQu7pi:-NqsMX7N-_aZMrfQu7pj:bxt7yz4) ([raw file](https://github.com/RobotLocomotion/drake/blob/a972b2cef27c2bba067c0d2319a73d547811e5f8/tools/wheel/image/setup.py#L46)):*
> ```Python
> # do to force the wheel to not be platform-agnostic. We need this because
> # trying to build an extension module with no sources is not reliable.
> with open('dummy.c', 'wt', encoding='ascii') as f:
> ```

This seems unnecessarily bespoke.  Why make us ponder whether the string literal has any utf-8 or not?  The hygiene we want is "UTF 8 everywhere".  Anything not marked as utf-8 should be viewed with extreme suspicion.



_Suggestion:_
```Python
encoding='utf-8'
```

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20640)* status: 1 unresolved discussion

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMAkL5HAthJllQLCf:-Nr2-sGu3gvQijZfqbsh:b-m6coyf) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Per https://peps.python.org/pep-0686/ and my related experience, the canonical string for `encoding=` in Python should be `"utf-8"` not `"utf8"`.  Please adjust throughout the PR.
</blockquote></details>

Done.

Interestingly, [codecs](https://docs.python.org/3/library/codecs.html#standard-encodings) seems to imply that `utf_8` is the "canonical" spelling, though `utf-8` seems most prevalent in practice (I can't say I've ever encountered the form with `_`!) and certainly within Drake. There are a number of other uses of `utf8` however that I guess we should fix (not in this PR). I wonder if this could be tool-enforced?

___
___
*[`tools/wheel/image/setup.py` line 46 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMX7N-_aZMrfQu7pi:-Nr21TPz2R6QZu3S3_Af:b-xqd3l2) ([raw file](https://github.com/RobotLocomotion/drake/blob/a972b2cef27c2bba067c0d2319a73d547811e5f8/tools/wheel/image/setup.py#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This seems unnecessarily bespoke.  Why make us ponder whether the string literal has any utf-8 or not?  The hygiene we want is "UTF 8 everywhere".  Anything not marked as utf-8 should be viewed with extreme suspicion.
</blockquote></details>

In point of fact, it *is* ASCII, though I suppose we trust the compiler to not choke on UTF-8?

Alternatively, what about:
```python
with open('dummy.c', 'wb') as f:
  f.write(b'void not_used() {}')
```

**Comment by mwoehlke-kitware** [comment]
*[`tools/wheel/image/setup.py` line 46 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMX7N-_aZMrfQu7pi:-Nr22C3ZBsLDmEyKTAmN:b-85qqko) ([raw file](https://github.com/RobotLocomotion/drake/blob/a972b2cef27c2bba067c0d2319a73d547811e5f8/tools/wheel/image/setup.py#L46)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

In point of fact, it *is* ASCII, though I suppose we trust the compiler to not choke on UTF-8?

Alternatively, what about:
```python
with open('dummy.c', 'wb') as f:
  f.write(b'void not_used() {}')
```
</blockquote></details>

I'd also be more suspicious except that `ascii` is a strict subset of `utf-8`. Thus, it isn't so much "not UTF-8" as "UTF-8 but *also* ASCII" (which is even "safer" than UTF-8 if we know we can use it).

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20640)* status: 1 unresolved discussion

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMAkL5HAthJllQLCf:-Nr242cK8YyEbR1wIP-I:b4s8qmn) (no related file):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Done.

Interestingly, [codecs](https://docs.python.org/3/library/codecs.html#standard-encodings) seems to imply that `utf_8` is the "canonical" spelling, though `utf-8` seems most prevalent in practice (I can't say I've ever encountered the form with `_`!) and certainly within Drake. There are a number of other uses of `utf8` however that I guess we should fix (not in this PR). I wonder if this could be tool-enforced?
</blockquote></details>

Perhaps it's plausible that `drakelint.py` could flag `encoding=["']utf8` with a linter complaint.  I also vaguely recall the upstream python tooling for linters talking about identifying missing (or bad?) encodings.  Maybe one of the existing ecosystem linters can flag it.

___
___
*[`tools/wheel/image/setup.py` line 46 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMX7N-_aZMrfQu7pi:-Nr24N189PH4fdbR56hT:b5tow9n) ([raw file](https://github.com/RobotLocomotion/drake/blob/a972b2cef27c2bba067c0d2319a73d547811e5f8/tools/wheel/image/setup.py#L46)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

I'd also be more suspicious except that `ascii` is a strict subset of `utf-8`. Thus, it isn't so much "not UTF-8" as "UTF-8 but *also* ASCII" (which is even "safer" than UTF-8 if we know we can use it).
</blockquote></details>

Yes, file content is ASCII and thus it's also valid UTF-8.

The point is that Drake code should be _boring_ w.r.t. encoding.  Boring means `encoding='utf-8'` plastered everywhere without ever giving it any conscious thought.  Breaking that pattern requires extra thought.  (The new developer needs to both check that it's _actually_ ASCII, and wonder what purpose the original developer was aiming for by breaking the pattern.)

For the same, using `bytes` would also be wrong, because it's not boring.

In short, Drake in effect and in my judgement has a rule of using `utf-8` everywhere.   It's just the coding style.  We should write it up in the style guide, but we haven't gotten to that yet.

**Review by mwoehlke-kitware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20640)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignee jwnimmer-tri(platform)

___
*[`tools/wheel/image/setup.py` line 46 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20640#-NqsMX7N-_aZMrfQu7pi:-Nr26Gm68Xaeci4zdEB6:baged7e) ([raw file](https://github.com/RobotLocomotion/drake/blob/a972b2cef27c2bba067c0d2319a73d547811e5f8/tools/wheel/image/setup.py#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, file content is ASCII and thus it's also valid UTF-8.

The point is that Drake code should be _boring_ w.r.t. encoding.  Boring means `encoding='utf-8'` plastered everywhere without ever giving it any conscious thought.  Breaking that pattern requires extra thought.  (The new developer needs to both check that it's _actually_ ASCII, and wonder what purpose the original developer was aiming for by breaking the pattern.)

For the same, using `bytes` would also be wrong, because it's not boring.

In short, Drake in effect and in my judgement has a rule of using `utf-8` everywhere.   It's just the coding style.  We should write it up in the style guide, but we haven't gotten to that yet.
</blockquote></details>

...which means we're relying on the compiler, *as invoked by `setuptools`* (not by us), to understand UTF-8.

The conservative in me would rather not do that.
The pragmatist doesn't want to argue about this.
